pkg/kubectl/cmd/run.go

```diff
@@ -146,10 +147,27 @@ func Run(f *cmdutil.Factory, cmdIn io.Reader, cmdOut, cmdErr io.Writer, cmd *cob
 
 	generatorName := cmdutil.GetFlagString(cmd, "generator")
 	if len(generatorName) == 0 {
+		client, err := f.Client()
+		if err != nil {
+			return err
+		}
+		groupList, err := client.Discovery().ServerGroups()
+		if err != nil {
+			// this cover the cases where old servers do not expose discovery
+			groupList = nil
+		}
 		if restartPolicy == api.RestartPolicyAlways {
-			generatorName = "deployment/v1beta1"
+			if contains(groupList, "extensions/v1beta1") {
+				generatorName = "deployment/v1beta1"
```

Comment:
Just because you have the group doesn't mean you enabled the object.  As I recall, we had levels with different resources enabled by default.  How about checking both?